adagrams/game.py

```diff
@@ -1,11 +1,186 @@
+
 def draw_letters():
-    pass
+    pool = []
+    for letter in letter_quant_dict:
+        quant = letter_quant_dict[letter]
+        while quant > 0:
+            pool.append(letter)
+            quant -= 1
+    hand = []
+    draw = 10
+    while draw > 0:
+        index = random.randint(0, len(pool)-1)
+        hand.append(pool[index])
+        pool.pop(index)
+        draw -= 1
+    return hand
+
+# def uses_available_letters(word, letter_bank):
+#     hand = letter_bank
+#     for letter in word.upper():
+#         if not letter in hand:
+#             return False
+#         hand.remove(letter)
+#     return True
 
 def uses_available_letters(word, letter_bank):
-    pass
+    #I dont understand why we woulndt want to update the letter_bank but ok
+
+    #build letter_bank_dict
+    letter_bank_dict = {}
+    for letter in letter_bank:
+        if letter in letter_bank_dict:
+            letter_bank_dict[letter] += 1
+        else:
+            letter_bank_dict[letter] = 1
+
+    #build hand_dict
+    hand_dict = {}
+    for letter in word.upper():
+        if letter in hand_dict:
+            hand_dict[letter] += 1
+        else:
+            hand_dict[letter] = 1
+    
+    #compare two dictionaries
+    for letter in hand_dict:
+        if letter not in letter_bank_dict:
+            return False
+        elif hand_dict[letter] > letter_bank_dict[letter]:
+            return False
+    return True
+
 
 def score_word(word):
-    pass
+    word_score = 0
+    if len(word) >= 7:
+        word_score += 8
+    for letter in word.upper():
+        letter_score = letter_score_dict[letter]
+        word_score += letter_score
+    return word_score
 
+# def get_highest_word_score(word_list):
+#     word_score_dict = {} 
+#     highest_score_yet = 0
+#     highest_word_yet = ""
+#     highest_len_yet = 0
+#     for word in word_list:
+#         word_score_dict[word] = score_word(word)
+#     for word, score in word_score_dict.items():
+#         if score > highest_score_yet:
+#             highest_score_yet = score
+#             highest_word_yet = word
+#             highest_len_yet = len(word)
+#         if score == highest_score_yet:
+#             if len(word) == 10:
+#                 highest_score_yet = score
+#                 highest_word_yet = word
+#                 highest_len_yet = len(word)
+#                 winning_tuple = (highest_word_yet, highest_score_yet)
+#             elif len(word) < highest_len_yet:
+#                 highest_score_yet = score
+#                 highest_word_yet = word
+#                 highest_len_yet = len(word)
+#     winning_tuple = (highest_word_yet, highest_score_yet)
+#     return winning_tuple
+    
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    highest_word_score = 0
+    highest_score_words = []
+    #find the highest score
+    for word in word_list:
+        if score_word(word) > highest_word_score:
+            highest_word_score = score_word(word)
```

Comment:
Nice approch to finding the max score. Try to avoid calling `score_word` more than once for a word (maybe tally it up for all of them once in a dictionary). We know that calling `score_word` conceals an inner loop, so the fewer times we call it the better.

And since you've manually implemented max logic (and min logic later), try to use the `max`/`min` functions the next time you need this logic (check out Higher Order Functions in Learn for a preview!)